File: libs/python/edgecommons/messaging/message_builder.py
"""
Builder for creating Message instances with fluent API.

``build()`` is the single UNS identity stamping site (UNS-CANONICAL-DESIGN §1.4): an
explicit :meth:`MessageBuilder.with_identity` override wins; otherwise, when a config
service is present, the component's resolved identity
(``ConfigManager.get_component_identity()``) is stamped with the per-message instance
token (:meth:`MessageBuilder.with_instance`, default ``"main"``); with neither,
``identity`` stays ``None`` (bootstrap/raw messages legally omit it).
"""
import json
from typing import TYPE_CHECKING, Optional

from edgecommons.messaging.identity import MessageIdentity
from edgecommons.messaging.proto import DEFAULT_OPAQUE_CONTENT_TYPE, MessageBodyCase, MessageBodySchema

if TYPE_CHECKING:
    from edgecommons.config.manager.config_manager import ConfigManager


class MessageBuilder:
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version
        self.correlation_id = None
        self.uuid = None
        self.timestamp = None
        self.timestamp_ms = None
        self.reply_to = None
        self.payload = None
        self.content_type = None
        self.content_encoding = None
        self.schema = None
        self.body_case = None
        self.config_service = None
        self.tags = None
        self.instance = None
        self.identity_override: Optional[MessageIdentity] = None

    @staticmethod
    def create(name: str, version: str) -> 'MessageBuilder':
        return MessageBuilder(name, version)

    @staticmethod
    def from_object(msg_contents) -> 'MessageBuilder':
        if isinstance(msg_contents, dict) and "header" in msg_contents:
            header = msg_contents["header"]
            name = header.get("name", "unknown")
            version = header.get("version", "1.0")

            builder = MessageBuilder(name, version)

            if "correlation_id" in header:
                builder.with_correlation_id(header["correlation_id"])
            if "uuid" in header:
                builder.with_uuid(header["uuid"])
            if "timestamp" in header:
                builder.with_timestamp(header["timestamp"])
            if "timestamp_ms" in header:
                builder.with_timestamp_ms(header["timestamp_ms"])
            if "reply_to" in header:
                builder.with_reply_to(header["reply_to"])
            if "identity" in msg_contents:
                # Lenient wire parse: a malformed identity yields None (no override).
                builder.with_identity(MessageIdentity.from_dict(msg_contents["identity"]))
            if "body" in msg_contents:
                builder.with_payload(msg_contents["body"])
            if "tags" in msg_contents:
                builder.with_tags(msg_contents["tags"])
            if "content_type" in msg_contents:
                builder.with_content_type(msg_contents["content_type"])
            if "content_encoding" in msg_contents:
                builder.with_content_encoding(msg_contents["content_encoding"])
            if "schema" in msg_contents:
                builder.with_schema(MessageBodySchema.from_dict(msg_contents["schema"]))
            if "body_case" in msg_contents:
                builder.with_body_case(MessageBodyCase[msg_contents["body_case"]])

            return builder
        else:
            # Raw message without header structure
            builder = MessageBuilder("raw", "1.0")
            builder.with_payload(msg_contents)
            return builder

    def with_correlation_id(self, correlation_id: str) -> 'MessageBuilder':
        self.correlation_id = correlation_id
        return self

    def with_payload(self, payload) -> 'MessageBuilder':
        self.payload = payload
        if isinstance(payload, (bytes, bytearray)) and self.body_case is None:
            self.body_case = MessageBodyCase.OPAQUE
            if self.content_type is None:
                self.content_type = DEFAULT_OPAQUE_CONTENT_TYPE
        return self

    def with_structured_payload(self, payload) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.STRUCTURED
        return self

    def with_structured_body(self, body) -> 'MessageBuilder':
        return self.with_structured_payload(body)

    def with_southbound_signal_update(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.SOUTHBOUND_SIGNAL_UPDATE
        return self

    def with_state_update(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.STATE_UPDATE
        return self

    def with_config_update(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.CONFIG_UPDATE
        return self

    def with_metric_update(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.METRIC_UPDATE
        return self

    def with_event(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.EVENT
        return self

    def with_command(self, payload: dict) -> 'MessageBuilder':
        self.payload = payload
        self.body_case = MessageBodyCase.COMMAND
        return self

    def with_opaque_payload(self, payload: bytes, content_type: str = DEFAULT_OPAQUE_CONTENT_TYPE) -> 'MessageBuilder':
        self.payload = None if payload is None else bytes(payload)
        self.body_case = MessageBodyCase.OPAQUE
        self.content_type = content_type if content_type is not None else DEFAULT_OPAQUE_CONTENT_TYPE
        return self

    def with_opaque_body(self, payload: bytes, content_type: str = DEFAULT_OPAQUE_CONTENT_TYPE) -> 'MessageBuilder':
        return self.with_opaque_payload(payload, content_type)

    def with_content_type(self, content_type: str) -> 'MessageBuilder':
        self.content_type = content_type
        return self

    def with_content_encoding(self, content_encoding: str) -> 'MessageBuilder':
        self.content_encoding = content_encoding
        return self

    def with_schema(self, schema: Optional[MessageBodySchema]) -> 'MessageBuilder':
        self.schema = schema
        return self

    def with_body_case(self, body_case) -> 'MessageBuilder':
        self.body_case = body_case if isinstance(body_case, MessageBodyCase) else MessageBodyCase[body_case]
        return self

    def with_config(self, config_service: 'ConfigManager') -> 'MessageBuilder':
        self.config_service = config_service
        return self

    def with_tags(self, tags: dict) -> 'MessageBuilder':
        self.tags = tags
        return self

    def with_uuid(self, uuid: str) -> 'MessageBuilder':
        """Pins the header ``uuid`` instead of the generated random one — deterministic
        envelopes for tests and the cross-language ``uns-test-vectors`` golden
        envelopes (D-U13)."""
        self.uuid = uuid
        return self

    def with_timestamp(self, timestamp: str) -> 'MessageBuilder':
        """Pins the header ``timestamp`` instead of the generated "now" — deterministic
        envelopes for tests and the cross-language ``uns-test-vectors`` golden
        envelopes (D-U13)."""
        self.timestamp = timestamp
        return self

    def with_timestamp_ms(self, timestamp_ms: int) -> 'MessageBuilder':
        self.timestamp_ms = int(timestamp_ms)
        return self

    def with_reply_to(self, reply_to: str) -> 'MessageBuilder':
        self.reply_to = reply_to
        return self

    def with_instance(self, instance: str) -> 'MessageBuilder':
        """Sets the per-message instance token stamped into the identity element
        (default ``"main"``). Only takes effect when an identity is stamped (a config
        service is present; an explicit identity override is stamped verbatim).

        :raises ValueError: if ``instance`` is ``None`` or empty
        """
        if not instance:
            raise ValueError("instance must be non-empty")
        self.instance = instance
        return self

    def with_identity(self, identity: Optional[MessageIdentity]) -> 'MessageBuilder':
        """Sets an explicit identity override (tests, conformance vectors, relays).
        Wins over the config-resolved identity and is stamped verbatim (the
        :meth:`with_instance` token is not applied to an override)."""
        self.identity_override = identity
        return self

    def build(self):
        from edgecommons.messaging.message import Message, MessageHeader, MessageTags

        message = Message()
        message.header = MessageHeader(self.name, self.version, self.correlation_id,
                                       self.timestamp, self.uuid, self.reply_to,
                                       self.timestamp_ms)

        if self.tags is not None:
            message.tags = MessageTags.from_dict(self.tags)
        elif self.config_service is not None:
            message.tags = MessageTags.from_config(self.config_service)

        # The single identity stamping site (§1.4): explicit override > config-resolved
        # component identity (+ per-message instance token) > none (bootstrap/raw cases
        # stay valid).
        if self.identity_override is not None:
            message.identity = self.identity_override
        elif self.config_service is not None:
            component_identity = self.config_service.get_component_identity()
            if component_identity is not None:
                message.identity = component_identity.with_instance(
                    self.instance if self.instance else MessageIdentity.DEFAULT_INSTANCE
                )

        if isinstance(self.payload, str):
            try:
                message.body = json.loads(self.payload)
            except json.JSONDecodeError:
                message.body = self.payload
        else:
            message.body = self.payload
        message.content_type = self.content_type
        message.content_encoding = self.content_encoding
        message.schema = self.schema
        message.body_case = self.body_case

        return message

File: libs/java/src/main/java/com/mbreissi/edgecommons/messaging/MessageBuilder.java
package com.mbreissi.edgecommons.messaging;

import com.mbreissi.edgecommons.config.ConfigManager;
import com.mbreissi.edgecommons.messaging.proto.MessageBodyCase;
import com.mbreissi.edgecommons.messaging.proto.MessageBodySchema;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;

/**
 * Builder for creating Message instances with fluent API.
 *
 * <p>{@link #build()} is the single UNS identity stamping site: an explicit
 * {@link #withIdentity(MessageIdentity)} override wins; otherwise, when a config service is
 * present, the component's resolved identity ({@code ConfigManager.getComponentIdentity()}) is
 * stamped with the per-message instance token ({@link #withInstance(String)}, default
 * {@value MessageIdentity#DEFAULT_INSTANCE}); with neither, {@code identity} stays {@code null}
 * (bootstrap/raw messages legally omit it).
 */
public class MessageBuilder {
    private String name;
    private String version;
    private String correlationId;
    private String uuid;
    private String timestamp;
    private Long timestampMs;
    private String replyTo;
    private Object payload;
    private String contentType;
    private String contentEncoding;
    private MessageBodySchema schema;
    private MessageBodyCase bodyCase;
    private MessageTags tagsOverride;
    private ConfigManager configService;
    private String instance;
    private MessageIdentity identityOverride;

    private MessageBuilder(String name, String version) {
        this.name = name;
        this.version = version;
    }

    public static MessageBuilder create(String name, String version) {
        return new MessageBuilder(name, version);
    }

    public static Message fromObject(Object msgContents) {
        Message retVal = new Message();
        if (msgContents instanceof JsonObject msgJsonObj)
        {
            if (msgJsonObj.has("header"))
            {
                retVal.header = MessageHeader.fromDict(msgJsonObj.getAsJsonObject("header"));
            }
            if (msgJsonObj.has("identity"))
            {
                retVal.identity = Message.parseIdentity(msgJsonObj.get("identity"));
            }
            if (msgJsonObj.has("tags"))
            {
                retVal.tags = MessageTags.fromDict(msgJsonObj.getAsJsonObject("tags"));
            }
            if (msgJsonObj.has("content_type"))
            {
                retVal.contentType = msgJsonObj.get("content_type").getAsString();
            }
            if (msgJsonObj.has("content_encoding"))
            {
                retVal.contentEncoding = msgJsonObj.get("content_encoding").getAsString();
            }
            if (msgJsonObj.has("schema"))
            {
                retVal.schema = MessageBodySchema.fromDict(msgJsonObj.getAsJsonObject("schema"));
            }
            if (msgJsonObj.has("body"))
            {
                retVal.body = msgJsonObj.get("body");
            }
            if (!(msgJsonObj.has("header") || msgJsonObj.has("identity")
                    || msgJsonObj.has("tags") || msgJsonObj.has("body")))
            {
                retVal.raw = msgJsonObj;
            }
        }
        else
        {
            retVal.raw = msgContents;
        }
        return retVal;
    }



    public MessageBuilder withCorrelationId(String correlationId) {
        this.correlationId = correlationId;
        return this;
    }

    /**
     * Pins the header {@code uuid} instead of the generated random one — deterministic envelopes
     * for tests and the cross-language {@code uns-test-vectors} golden envelopes (D-U13).
     *
     * @param uuid the header uuid to stamp verbatim
     */
    public MessageBuilder withUuid(String uuid) {
        this.uuid = uuid;
        return this;
    }

    /**
     * Pins the header {@code timestamp} instead of the generated "now" — deterministic envelopes
     * for tests and the cross-language {@code uns-test-vectors} golden envelopes (D-U13).
     *
     * @param timestamp the header timestamp to stamp verbatim (ISO-8601 instant by convention)
     */
    public MessageBuilder withTimestamp(String timestamp) {
        this.timestamp = timestamp;
        return this;
    }

    public MessageBuilder withTimestampMs(long timestampMs) {
        this.timestampMs = timestampMs;
        return this;
    }

    public MessageBuilder withReplyTo(String replyTo) {
        this.replyTo = replyTo;
        return this;
    }

    public MessageBuilder withPayload(Object payload) {
        this.payload = payload;
        if (payload instanceof byte[] && this.bodyCase == null) {
            this.bodyCase = MessageBodyCase.OPAQUE;
            if (this.contentType == null) {
                this.contentType = "application/octet-stream";
            }
        }
        return this;
    }

    public MessageBuilder withStructuredPayload(Object payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.STRUCTURED;
        return this;
    }

    public MessageBuilder withStructuredBody(Object body) {
        return withStructuredPayload(body);
    }

    public MessageBuilder withSouthboundSignalUpdate(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.SOUTHBOUND_SIGNAL_UPDATE;
        return this;
    }

    public MessageBuilder withStateUpdate(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.STATE_UPDATE;
        return this;
    }

    public MessageBuilder withConfigUpdate(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.CONFIG_UPDATE;
        return this;
    }

    public MessageBuilder withMetricUpdate(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.METRIC_UPDATE;
        return this;
    }

    public MessageBuilder withEvent(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.EVENT;
        return this;
    }

    public MessageBuilder withCommand(JsonObject payload) {
        this.payload = payload;
        this.bodyCase = MessageBodyCase.COMMAND;
        return this;
    }

    public MessageBuilder withOpaquePayload(byte[] payload) {
        return withOpaquePayload(payload, "application/octet-stream");
    }

    public MessageBuilder withOpaquePayload(byte[] payload, String contentType) {
        this.payload = payload == null ? null : payload.clone();
        this.bodyCase = MessageBodyCase.OPAQUE;
        this.contentType = contentType != null ? contentType : "application/octet-stream";
        return this;
    }

    public MessageBuilder withOpaqueBody(byte[] payload) {
        return withOpaquePayload(payload);
    }

    public MessageBuilder withOpaqueBody(byte[] payload, String contentType) {
        return withOpaquePayload(payload, contentType);
    }

    public MessageBuilder withContentType(String contentType) {
        this.contentType = contentType;
        return this;
    }

    public MessageBuilder withContentEncoding(String contentEncoding) {
        this.contentEncoding = contentEncoding;
        return this;
    }

    public MessageBuilder withSchema(MessageBodySchema schema) {
        this.schema = schema;
        return this;
    }

    public MessageBuilder withBodyCase(MessageBodyCase bodyCase) {
        this.bodyCase = bodyCase;
        return this;
    }

    public MessageBuilder withTags(MessageTags tags) {
        this.tagsOverride = tags;
        return this;
    }

    public MessageBuilder withConfig(ConfigManager configService) {
        this.configService = configService;
        return this;
    }

    /**
     * Sets the per-message instance token stamped into the identity element. A {@code null}/empty
     * token means component scope (D‑U28: the identity carries no instance key). Only takes effect
     * when an identity is stamped (a config service is present or an explicit identity override
     * carries it).
     *
     * @param instance the instance token, or {@code null}/empty for component/global scope
     */
    public MessageBuilder withInstance(String instance) {
        this.instance = (instance == null || instance.isEmpty()) ? null : instance;
        return this;
    }

    /**
     * Sets an explicit identity override (tests, conformance vectors, relays). Wins over the
     * config-resolved identity and is stamped verbatim (the {@link #withInstance(String)} token
     * is not applied to an override).
     *
     * @param identity the identity to stamp
     */
    public MessageBuilder withIdentity(MessageIdentity identity) {
        this.identityOverride = identity;
        return this;
    }

    public Message build() {
        Message message = new Message();
        MessageHeaderBuilder headerBuilder = MessageHeaderBuilder.create(name, version);
        if (correlationId != null) {
            headerBuilder.withCorrelationId(correlationId);
        }
        if (uuid != null) {
            headerBuilder.withUuid(uuid);
        }
        if (timestamp != null) {
            headerBuilder.withTimestamp(timestamp);
        }
        if (timestampMs != null) {
            headerBuilder.withTimestampMs(timestampMs);
        }
        if (replyTo != null) {
            headerBuilder.withReplyTo(replyTo);
        }
        message.header = headerBuilder.build();
        if (tagsOverride != null) {
            message.tags = tagsOverride;
        } else if (configService != null) {
            message.tags = MessageTags.fromConfig(configService);
        }

        // The single identity stamping site: explicit override > config-resolved component
        // identity (+ per-message instance token) > none (bootstrap/raw cases stay valid).
        if (identityOverride != null) {
            message.identity = identityOverride;
        } else if (configService != null) {
            MessageIdentity componentIdentity = configService.getComponentIdentity();
            if (componentIdentity != null) {
                message.identity = componentIdentity.withInstance(instance);   // D‑U28: null ⇒ component scope
            }
        }

        if (payload instanceof String payloadStr) {
            try {
                Gson gson = new Gson();
                message.body = gson.fromJson(payloadStr, Object.class);
            } catch (JsonSyntaxException e) {
                message.body = payloadStr;
            }
        } else {
            message.body = payload;
        }
        message.contentType = contentType;
        message.contentEncoding = contentEncoding;
        message.schema = schema;
        message.bodyCase = bodyCase;

        return message;
    }
}
